src/residuals.cc

```diff
@@ -0,0 +1,155 @@
+                             const bcf1_t *gl_call,
+                             std::string &ynode) {
+    Status s;
+    YAML::Emitter out;
+
+    out << YAML::BeginMap;
+
+    // write the original records
+    out << YAML::Key << "input";
+    out << YAML::BeginMap;
+    out << YAML::Key << "body";
+    out << YAML::BeginSeq;
+    shared_ptr<const set<string>> samples2, datasets;
+    vector<unique_ptr<RangeBCFIterator>> iterators;
+    S(data_.sampleset_range(cache_, sampleset_, site.pos,
+                           samples2, datasets, iterators));
+    assert(samples_.size() == samples2->size());
+
+    // for each pertinent dataset
+    for (const auto& dataset : *datasets) {
+        // load BCF records overlapping the site by "merging" the iterators
+        shared_ptr<const bcf_hdr_t> dataset_header;
+        vector<vector<shared_ptr<bcf1_t>>> recordss(iterators.size());
+        vector<shared_ptr<bcf1_t>> records;
+
+        for (const auto& iter : iterators) {
+            string this_dataset;
+            vector<shared_ptr<bcf1_t>> these_records;
+            S(iter->next(this_dataset, dataset_header, these_records));
+            if (dataset != this_dataset) {
+                return Status::Failure("genotype_site: iterator returned unexpected dataset",
+                                       this_dataset + " instead of " + dataset);
+            }
+            records.insert(records.end(), these_records.begin(), these_records.end());
+        }
+
+        for (auto rec : records) {
+            out << YAML::BeginMap;
+            out << YAML::Key << dataset;
+            string samples = samples_from_dataset(dataset_header.get());
+            out << YAML::Literal << samples + "\n" + *(bcf1_to_string(dataset_header.get(), rec.get()));
+            out << YAML::EndMap;
+        }
+    }
+    out << YAML::EndSeq;
+    out << YAML::EndMap;
+
+    // write the unified site
+    out << YAML::Key << "unified_site";
+    const auto& contigs = cache_.contigs();
+    S(site.yaml(contigs, out));
+
+
+    // write the output bcf, the calls that GLnexus made.
+    out << YAML::Key << "output_vcf";
+    out << YAML::Literal << *(bcf1_to_string(gl_hdr, gl_call));
+
+    out << YAML::EndMap;
+
+    // Emit YAML format
+    ynode = string(out.c_str());
+    return Status::OK();
+}
+
+// destructor
+ResidualsFile::~ResidualsFile() {
+    ofs_.close();
+}
+
+
+static bool is_file_exist(const string &fileName) {
+    std::ifstream infile(fileName);
+    return infile.good();
+}
+
+Status ResidualsFile::Open(std::string filename,
+                                  std::unique_ptr<ResidualsFile> &ans) {
+    ans = make_unique<ResidualsFile>(filename);
+
+    if (is_file_exist(filename)) {
+        int rc = remove(filename.c_str());
+        if (rc != 0)
+            return Status::Invalid("Residuals file cannot be removed ", filename);
+    }
+    ans->ofs_.open(filename, std::ofstream::out | std::ofstream::app);
```

Comment:
I first erase the file, so that in case of a later failure, I will get an empty file, and not the results of a previous run. I will add a comment explaining this.